### PR TITLE
EOS-18824 : Validate keys extracted for each phase from each phase of provisioning

### PR DESCRIPTION
**Pull request by shahidSTX**
Validate keys extracted from template files against confstore keys

Files modified :
	scripts/provisioning/postinstallcmd.py

Testing :
       In progress.

Signed-off-by: Mohammed Shahid <mohammed.shahid@seagate.com>

### CODE REVIEW

File: scripts/provisioning/postinstallcmd.py [hunk 1]
```diff
@@ -68,3 +70,21 @@ def validate_pre_requisites(self,
       sys.stderr.write('ERROR: post_install validations failed.\n')
       sys.stderr.write(f"{e}, config:{self._preqs_conf_file}\n")
       raise e
+
+  def  extract_keys_from_template(self):
+    """Extract all keys in template."""
+    index = "postinstall_s3keys_index"
+    filename = "/opt/seagate/cortx/s3/conf/s3.post_install.tmpl.1-node.sample"
```
Comment (nileshgovande) [inline scripts/provisioning/postinstallcmd.py:77]: take file from config commandline

File: scripts/provisioning/setupcmd.py [hunk 2]
```diff
@@ -157,6 +157,39 @@ def phase_prereqs_validate(self, phase_name: str):
     except Exception as e:
       raise S3PROVError(f'ERROR: {phase_name} prereqs validations failed, exception: {e} \n')
 
+  def phase_keys_validate(self, url_file: str, phase_name: str):
+    """Validate keys of each phase, extract from template file and compare with s3_prov_config."""
+    phase_name = phase_name.upper()
+    template_keys_confstore = S3CortxConfStore(url_file, 'keys_template_index')
```
Comment (nileshgovande) [inline scripts/provisioning/setupcmd.py:163]: arguement_file_index

File: scripts/provisioning/preparecmd.py [hunk 3]
```diff
@@ -37,3 +37,4 @@ def process(self):
     """Main processing function."""
     sys.stdout.write(f"Processing {self.name} {self.url}\n")
     self.phase_prereqs_validate(self.name)
+    self.phase_keys_validate(self.url, self.name)
```
[the remark this answers is not in the document]
Comment (shahidSTX) [reply scripts/provisioning/preparecmd.py:40]: Done.

File: scripts/provisioning/postinstallcmd.py (hunk 1 above)
Comment (shahidSTX) [reply scripts/provisioning/postinstallcmd.py]: Done.

File: scripts/provisioning/setupcmd.py (hunk 2 above)
Comment (shahidSTX) [reply scripts/provisioning/setupcmd.py]: Done

File: scripts/provisioning/setupcmd.py [hunk 4]
```diff
@@ -167,6 +168,98 @@ def phase_prereqs_validate(self, phase_name: str):
...
+      # array entries. The result will be stored in yardstick
+      # list which will be complete and will be used to verify
+      # keys required for each phase.
+      yardstick_list = []
+      for key in arg_keys_list:
+        if ((key.find('[') != -1) and (key.find(']') != -1)):
+          storage_set = self.get_confvalue(key)
+          for set_key in storage_set:
+            key = key + ">" + set_key
+        yardstick_list.append(key)
```
Comment (shahidSTX) [inline scripts/provisioning/setupcmd.py:190]: Change yardstick to argument.

**Review by nileshgovande — APPROVED** [review]

File: scripts/provisioning/setupcmd.py (hunk 4 above)
Comment (shahidSTX) [reply scripts/provisioning/setupcmd.py]: Done.

File: scripts/provisioning/setupcmd.py [hunk 5]
```diff
@@ -157,6 +157,39 @@ def phase_prereqs_validate(self, phase_name: str):
...
+        new_url_keys_list.append(key)
+      prov_keys_list = self._s3_confkeys_store.get_all_keys()
+      phase_keys_list = []
+      for key in prov_keys_list:
+        if key.find(phase_name) != 0:
+          continue
+        newkey = self.get_confkey(key)
+        if newkey is None:
+          continue
+        if newkey.find("server_node") == 0:
```
[the remark this answers is not in the document]
Comment (shahidSTX) [reply scripts/provisioning/setupcmd.py]: Removed this code.

**Comment by shahidSTX** [comment]
Update : mini-provisioning is failing here because the dependent PR #805 is not merged. We expect the base file (s3_prov_config.yaml) to be updated with hierarchy as a consumable for this PR.

**Review by AjinkyaDhumal — APPROVED** [review]

File: scripts/provisioning/setupcmd.py [hunk 6]
```diff
@@ -175,6 +175,143 @@ def phase_prereqs_validate(self, phase_name: str):
...
+
+      prov_keys_list = self._s3_confkeys_store.get_all_keys()
+
+      # Below hard-coded key is used to set the value
+      # of 'storage_set_count'. This will be used to
+      # set the array size of 'storage_set' if it is
+      # found in the key string.
+      site_count_key = "cluster>cluster-id>site>storage_set_count"
+      if not self.cluster_id is None:
+        site_count_key = site_count_key.replace("cluster-id", self.cluster_id)
```
Comment (shahidSTX) [inline scripts/provisioning/setupcmd.py:237]: set_count*
Fetch this, cluster-id and machine-id on top.

Comment (shahidSTX) [reply scripts/provisioning/setupcmd.py]: Done.

File: scripts/provisioning/setupcmd.py [hunk 7]
```diff
@@ -175,6 +176,170 @@ def phase_prereqs_validate(self, phase_name: str):
...
+    storage_set_count_key = "cluster>cluster-id>site>storage_set_count"
+    if self.cluster_id is not None:
+      storage_set_count_key = storage_set_count_key.replace("cluster-id", cluster_id_val)
+    storage_set_count_str = self.get_confvalue(storage_set_count_key)
+    if storage_set_count_str is not None:
+      storage_set_val = int(storage_set_count_str)
+    else:
+      storage_set_val = 0
+    # Set phase name to upper case required for inheritance
+    phase_name = phase_name.upper()
```
Comment (shahidSTX) [inline scripts/provisioning/setupcmd.py:268]: Add try after here.

Comment (shahidSTX) [reply scripts/provisioning/setupcmd.py:267]: Done.

**Review by nileshgovande — APPROVED** [review]

**Review by vimalraghwani20 — APPROVED** [review]

**Comment by shahidSTX** [comment]
Re-triggered [mini-provisioning](http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/852/) and [pre-merge test](http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Component-Test/job/S3Server-PreMerge-Test/807/). Both jobs passed.

**Pull request merged**